src/callback_handler/mod.rs

```diff
@@ -107,6 +30,28 @@ pub struct CallbackHandler {
     shutdown_channel: oneshot::Receiver<()>,
 }
 
+macro_rules! handle_callback {
+    ($req:expr, $log_value: expr, $log_msg: expr, $code:block) => {{
+        let response = { $code }
+            .await
+            .map(|response| {
+                debug!(
+                    value = ?$log_value,
+                    cached = response.was_cached,
+                    $log_msg,
+                );
+                let payload = serde_json::to_vec(&response.value)
+                    .map_err(|e| anyhow!("error serializing payload: {e:?}"))?;
+                Ok(CallbackResponse { payload })
+            })
+            .and_then(|r| r);
```

Comment:
this waits until the response Future has succeeded and returns it. If not, we issue the warn, correct?
We are tracing it as warn because that's what we did. Should it be error?

Reply:
It waits for the future to complete. The response returns a `Result<something>`. If the response is `Ok` we enter the `map(|response|)` block, which converts the response object into a json string.

However, the transformation can fail too. Hence the `map` block returns a `Result<Result<Vec<u8>>>`. The `and_then` block is executed **only if** the original response was `Ok`, it's purpose is to flatten the `Result<Result<>>` to be a `Result<Vec<u8>>`.

Long story short: this is convoluted, but no error is lost during the process. There's a rust nightly feature called [`Result::flatten`](https://doc.rust-lang.org/std/result/enum.Result.html#method.flatten) which would be more explicit, but this works in the same way

Reply:
I see it now. Thanks!